src/uct/tcp/tcp_ep.c

```diff
@@ -347,11 +340,31 @@ void uct_tcp_ep_destroy(uct_ep_h tl_ep)
     uct_tcp_ep_t *ep       = ucs_derived_of(tl_ep, uct_tcp_ep_t);
     uct_tcp_iface_t *iface = ucs_derived_of(ep->super.super.iface,
                                             uct_tcp_iface_t);
+    ucs_status_t status;
 
-    if (/* EPs that are connected as CONNECT_TO_EP have to be full duplex */
-        !(ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP) &&
-        (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED) &&
-        ucs_test_all_flags(ep->flags, UCT_TCP_EP_CTX_CAPS)) {
+    if ((ep->flags & UCT_TCP_EP_FLAG_CONNECT_TO_EP) &&
+        (ep->conn_state == UCT_TCP_EP_CONN_STATE_CONNECTED)) {
+        ucs_assert(!(ep->flags & UCT_TCP_EP_FLAG_ON_PTR_MAP));
+
+        ucs_assert(ep->flags & UCT_TCP_EP_FLAG_CTX_TYPE_TX);
+        uct_tcp_ep_remove_ctx_cap(ep, UCT_TCP_EP_FLAG_CTX_TYPE_TX);
+
+        if (!uct_tcp_cm_ep_accept_conn(ep)) {
+            /* If this is EP that is responsible for connecting to the peer,
+             * need to close the connection greacefully syncing with peer by
+             * sending CONN_CLOSE_REQ (if nothing is scheduled in TX) */
```

Comment:
why do we need this?
can't we just close the socket?

Reply:
We couldn't just close a connection, since the peer will receive an event about a dropped connection and will start the error handling procedure (so, multiple client-server tests failed due to this)
I have an example that reproduces this issue very well:
1. CM phase was done for two TCP_SOCKCM lanes;
2. Server sent `WIREUP_MSG/PRE_REQ` to client.
3. Client handled `WIREUP_MSG/PRE_REQ`, sent `WIREUP_MSG/REQ` back to sever;
4. Server handled `WIREUP_MSG/REQ`, reconfigured EP and use other lanes than it was after CM phase (i.e. UCT/TCP EP created during CM phase - is AUX EP in CM_WIIREUP_EP), sent `WIREUP_MSG/REP` back to client;
5. Client handled `WIREUP_MSG/REP`, reconfigured EP and use other lanes than it was after CM phase (i.e. UCT/TCP EP created during CM phase - is AUX EP in CM_WIIREUP_EP), make UCP EP as `REMOTE_CONNECTED` (i.e. it changes all lanes from WIREUP_EP to the real UCT EPs and closes AUX_EP from CM_WIREUP_EP as well), sent `WIREUP_MSG/ACK` back to server using AM lane;
6. Server detected error on CM_WIREUP_EP/AUX_EP, since `WIREUP_MSG/ACK` wasn't received yet (i.e. UCP EP was not marked as `REMOTE_CONNECTED`) and the connection was destroyed by the client. It calls error handling callback and cleanup lanes on UCP EP.

Reply:
in (6): can we ignore errors on aux_ep if we already connected real eps?
what would happen if we had RC transport instead of TCP, would it work, or require some "fixes" in RC as well?

Reply:
> in (6): can we ignore errors on aux_ep if we already connected real eps?

but they are not `REMOTE_CONNECTED`. do you mean rely on that fact that `WIREUP_MSG/REP` was sent from server and it means that we are just waiting `ACK`?
But what if some real issue happens on a client-side before it is managed to send `WIREUP_MSG/ACK`? it will be ignored and UCX will hang then

> what would happen if we had RC transport instead of TCP, would it work, or require some "fixes" in RC as well?

looks like either "TCP is too slow to receive `WIREUP_MSG/ACK` on a server, close the CM_WIREUP_EP/AUX_EP and not detect an error" or "RC is too slow to get an error on server side about a closed connection (RNR NAK timeout?)"

Reply:
Anyway, if we will decide to ignore an error from AUX_EP in case of `WIREUP_MSG/REP` sent, but `WIREUP_MSG/ACK` is not received, we should have this fix at UCT/TCP level to support other possible scenarios that could happen at UCP level or UCT users.

Reply:
ignoring failure for `CM_WIREUP_EP/AUX_EP` was implemented when a server EP is about to receive ACK